File: public/public-persistence-api/src/main/java/ch/difty/scipamato/publ/persistence/api/NewStudyTopicService.java
package ch.difty.scipamato.publ.persistence.api;

import java.util.List;

import org.jetbrains.annotations.NotNull;

import ch.difty.scipamato.publ.entity.NewStudyPageLink;
import ch.difty.scipamato.publ.entity.NewStudyTopic;
import ch.difty.scipamato.publ.entity.Newsletter;

public interface NewStudyTopicService {

    /**
     * Finds the all {@link NewStudyTopic}s associated with the newest newsletter.
     * The Topics are translated into the language specified with the languageCode.
     *
     * @param languageCode
     *     the two character language code (used to translate the topics)
     * @return list of new study topics, will not be null.
     */
    @NotNull
    List<NewStudyTopic> findMostRecentNewStudyTopics(@NotNull String languageCode);

    /**
     * Finds the all {@link NewStudyTopic}s associated with the newsletter with the specified issue.
     * The Topics are translated into the language specified with the languageCode.
     *
     * @param issue
     *     the issue of the newsletter (e.g. 2018/06)
     * @param languageCode
     *     the two character language code (used to translate the topics)
     * @return list of new study topics, will not be null
     */
    @NotNull
    List<NewStudyTopic> findNewStudyTopicsForNewsletterIssue(@NotNull String issue, @NotNull String languageCode);

    /**
     * Returns a list with the most recent newsletters.
     *
     * @param newsletterCount
     *     the number of newsletters that shall be offered.
     * @param languageCode
     *     the two character languageCode (e.g. 'en')
     * @return a list of {@link Newsletter}s
     */
    @NotNull
    List<Newsletter> findArchivedNewsletters(int newsletterCount, @NotNull String languageCode);

    /**
     * Returns a list with links to display on the new study page.
     *
     * @param languageCode
     *     the two character languageCode (e.g. 'en')
     * @return a list of {@link NewStudyPageLink}s
     */
    @NotNull
    List<NewStudyPageLink> findNewStudyPageLinks(@NotNull String languageCode);
}

File: public/public-persistence-api/src/main/kotlin/ch/difty/scipamato/publ/persistence/api/NewStudyTopicService.kt
package ch.difty.scipamato.publ.persistence.api

import ch.difty.scipamato.publ.entity.NewStudyPageLink
import ch.difty.scipamato.publ.entity.NewStudyTopic
import ch.difty.scipamato.publ.entity.Newsletter

interface NewStudyTopicService {

    /**
     * Finds the all [NewStudyTopic]s associated with the newest newsletter.
     * The Topics are translated into the language specified with the provided [languageCode].
     */
    fun findMostRecentNewStudyTopics(languageCode: String): List<NewStudyTopic>

    /**
     * Finds the all [NewStudyTopic]s associated with the newsletter with the specified [issue] (e.g. 2018/06).
     * The Topics are translated into the language specified with the [languageCode].
     */
    fun findNewStudyTopicsForNewsletterIssue(issue: String, languageCode: String): List<NewStudyTopic>

    /**
     * Returns a list with the most recent newsletters.
     *
     * @param newsletterCount the number of newsletters that shall be offered.
     * @param languageCode the two character languageCode (e.g. 'en')
     */
    fun findArchivedNewsletters(newsletterCount: Int, languageCode: String): List<Newsletter>

    /**
     * Returns a list with links to display on the new study page, according to the provided [languageCode].
     */
    fun findNewStudyPageLinks(languageCode: String): List<NewStudyPageLink>
}
